compositor_pipeline/src/audio_mixer/internal_audio_mixer.rs

```diff
@@ -189,4 +60,184 @@ impl InternalAudioMixer {
+        let start_pts = samples_set.start_pts;
+        let (input_samples, samples_count) = self.merge_fill_input_samples(samples_set);
+
+        OutputSamplesSet(
+            self.outputs
+                .iter()
+                .map(|(output_id, output_info)| {
+                    let samples = mix(&input_samples, output_info, samples_count);
+                    (output_id.clone(), OutputSamples { samples, start_pts })
+                })
+                .collect(),
+        )
+    }
+
+    fn merge_fill_input_samples(
+        &self,
+        input_samples_set: InputSamplesSet,
+    ) -> (HashMap<InputId, Vec<(i16, i16)>>, usize) {
+        let samples_count = (input_samples_set
+            .end_pts
+            .saturating_sub(input_samples_set.start_pts)
+            .as_secs_f64()
+            * self.output_sample_rate as f64)
+            .round() as usize;
+
+        let input_samples = input_samples_set
+            .samples
+            .into_iter()
+            .map(|(input_id, input_batch)| {
+                let samples = Self::frame_input_samples(
+                    input_samples_set.start_pts,
+                    input_samples_set.end_pts,
+                    input_batch,
+                    self.output_sample_rate,
+                );
+
+                (input_id, samples)
+            })
+            .collect();
+        (input_samples, samples_count)
+    }
+
+    fn frame_input_samples(
+        start_pts: Duration,
+        end_pts: Duration,
+        samples: Vec<InputSamples>,
+        sample_rate: u32,
+    ) -> Vec<(i16, i16)> {
+        let mut samples_in_frame = Vec::new();
+
+        samples
+            .into_iter()
+            .fold(start_pts, |last_end_pts, input_samples| {
+                let time_since_last_end = input_samples
+                    .start_pts
+                    .saturating_sub(last_end_pts)
+                    .as_secs_f64();
+                let missing_samples = (time_since_last_end * sample_rate as f64).round() as usize;
+                if missing_samples > 1 {
+                    Self::push_missing_samples(&mut samples_in_frame, missing_samples);
+                };
+                // Pushing InputSamples in frame
+                let sample_time_before_frame = start_pts
+                    .saturating_sub(input_samples.start_pts)
+                    .as_secs_f64();
+                let start_index = (sample_time_before_frame * sample_rate as f64).round() as usize;
+
+                let sample_time_after_frame =
+                    input_samples.end_pts.saturating_sub(end_pts).as_secs_f64();
+                let samples_after_frame =
+                    (sample_time_after_frame * sample_rate as f64).round() as usize;
+                let end_index = input_samples.len() - samples_after_frame;
+
+                samples_in_frame.extend(input_samples.samples[start_index..end_index].iter());
+
+                input_samples.end_pts
+            });
```

Comment:
I think that it's better to check:
```RUST
// SOME_SMALL_POSITIVE_VAL is sth like 3 or 5
missing_samples > SOME_SMALL_POSITIVE_VAL 
```
rather than:
```RUST
missing_samples > 0 
```
There are only a few reasons for missing samples being >0 here:
1.  Compositor hasn't received/read some input samples. eg. compositor received/read samples [0ms, 20ms] and [40ms, 60ms] (lost packets, missing audio fragments, etc.)
2. Audio is correct, but some PTSs are a little off, due to converting between different formats and numerically unsafe calculation of missing samples (eg. converting from RTP range timestamps to Duration can introduce small errors in PTS values)

We should fill samples only in the 1st case.

I think that with values like `missing_samples = 1` or `missing_samples = 2` it's significantly more probable that it's the 2nd reason. Nobody sends packets with only a few audio samples or puts 1/44100s of audio into the MP4 segment.